options/posix/generic/arpa-inet.cpp

```diff
@@ -200,9 +201,89 @@ int inet_pton(int af, const char *__restrict src, void *__restrict dst) {
 			addr->s_addr = htonl(ip);
 			break;
 		}
-		case AF_INET6:
-			mlibc::infoLogger() << "inet_pton: ipv6 is not implemented!" << frg::endlog;
-			/* fallthrough */
+		case AF_INET6: {
+			if(strchr(src, '.') != nullptr) {
+				mlibc::infoLogger() << "inet_pton: ipv4 in ipv6 address is not supported" << frg::endlog;
+				return 0;
+			}
+
+			const char *orig_src = src;
+			char *abbrev = strstr(src, "::");
+			if(abbrev && strstr(abbrev + 2, "::") != nullptr) {
+				return 0;
+			}
+
+			uint16_t addr[8] = {};
+			int len = abbrev ? abbrev - src : strlen(src);
+
+			int addr_len = 0;
+			while(*src && (src - orig_src) < len) {
+				char *end;
+				long int value = strtol(src, &end, 16);
+
+				if(value > 0xffff) {
+					return 0;
+				}
+				if(*end != '\0' && *end != ':') {
+					return 0;
+				}
+
+				src = end + 1;
+
+				if(addr_len >= 8) {
+					return 0;
+				}
+
+				addr[addr_len++] = htons(value);
+			}
+
+			if(abbrev) {
+				uint16_t array[8] = {};
+				int array_len = 0;
+
+				src = abbrev + 2;
+				while(*src) {
```

Comment:
Why is this case basically copy pasted from the above one? Can they not be merged?